### PR TITLE
Rule S3400: Fix FP when using default interface implementation

### LINKED ISSUE #5498: Rule S3400: FP when using default interface implementation
### Description

S3400 raises an issue for default interface implementation. The method can be overridden so the behavior should be similar with virtual methods.

Source: https://community.sonarsource.com/t/csharpsquid-s3400-should-not-apply-to-default-interface-implementations/60075

### Repro steps

```
interface ISomeInterface {
    public bool IsCondition() => true; // this triggers the rule
}

class SomeClass implements ISomeInterface {
    public bool Value { get; set; }
    public bool IsCondition() => Value; // this is not a constant
}
```

### Expected behavior

The rule should not be raised.

### Actual behavior

The `IsCondition` method is considered noncompliant.

### Known workarounds

Suppress the raised warning.

### Related information

* C#/VB.NET Plugins version: https://github.com/SonarSource/sonar-dotnet/releases/tag/8.36.1.44192

### PR DESCRIPTION
**Pull request by costin-zaharia-sonarsource**
Fixes #5498

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodShouldNotOnlyReturnConstant.cs [hunk 1]
```diff
@@ -56,8 +56,8 @@ public sealed class MethodShouldNotOnlyReturnConstant : SonarDiagnosticAnalyzer
                     }
 
                     var methodSymbol = c.SemanticModel.GetDeclaredSymbol(methodDeclaration);
-
                     if (methodSymbol != null
+                        && !methodSymbol.ContainingType.IsInterface()
                         && methodSymbol.GetInterfaceMember() == null
                         && methodSymbol.GetOverriddenMember() == null)
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MethodShouldNotOnlyReturnConstant.cs:62]: Can be:
```
                    if (c.SemanticModel.GetDeclaredSymbol(methodDeclaration) is { } methodSymbol
                        && !methodSymbol.ContainingType.IsInterface()
                        && methodSymbol.GetInterfaceMember() == null
                        && methodSymbol.GetOverriddenMember() == null)
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/MethodShouldNotOnlyReturnConstant.cs [hunk 2]
```diff
@@ -56,8 +56,8 @@ public sealed class MethodShouldNotOnlyReturnConstant : SonarDiagnosticAnalyzer
...
                     {
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/MethodShouldNotOnlyReturnConstant.cs:63]: In this file     IsConstantExpression can be an arrow bodied method with
  
```
expression.RemoveParentheses() is LiteralExpressionSyntax noParenthesesExpression
&& !noParenthesesExpression.IsNullLiteral()
&& semanticModel.GetConstantValue(noParenthesesExpression).HasValue;
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ExceptionsNeedStandardConstructors.cs [hunk 3]
```diff
@@ -68,18 +69,12 @@ private static bool HasStandardConstructors(INamedTypeSymbol classSymbol)
...
 
-        private static bool IsMatchingConstructor(IMethodSymbol constructor, Accessibility accessibility,
-            KnownType[] expectedParameterTypes)
-        {
-            return constructor.DeclaredAccessibility == accessibility &&
-                   EnumerableExtensions.Equals(constructor.Parameters, expectedParameterTypes,
-                       (p1, p2) => p1.Type.Is(p2));
-        }
+        private static bool IsMatchingConstructor(IMethodSymbol constructor, Accessibility accessibility, KnownType[] expectedParameterTypes) =>
+            constructor.DeclaredAccessibility == accessibility &&
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ExceptionsNeedStandardConstructors.cs:76]: The && operator is in the wrong place.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ExceptionsNeedStandardConstructors.cs [hunk 4]
```diff
@@ -68,18 +69,12 @@ private static bool HasStandardConstructors(INamedTypeSymbol classSymbol)
...
+            EnumerableExtensions.Equals(constructor.Parameters, expectedParameterTypes,
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ExceptionsNeedStandardConstructors.cs:77]: The formatting of the arguments is not aligned with our coding style.

File: analyzers/src/SonarAnalyzer.Common/Extensions/ITypeSymbolExtensions.cs [hunk 5]
```diff
@@ -21,13 +21,12 @@
...
 
-namespace SonarAnalyzer.Helpers
+namespace SonarAnalyzer.Extensions
 {
-    internal static class TypeHelper
+    internal static class ITypeSymbolExtensions
     {
-        #region TypeKind
-
         public static bool IsInterface(this ITypeSymbol self) =>
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Extensions/ITypeSymbolExtensions.cs:30]: Not to repeat the logic I suggest calling Is() on self with the correct typekind for the 3 checks: IsStruct, IsInterface, IsClass

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
Looks good, I left a few comments. Two of those are related to the original FP fix, the others are part of the refactoring. I suggest to separate the 2nd and 3rd commits to a separate PR.
Also the coverage should be easy to bump for the extension methods.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ExceptionsNeedStandardConstructors.cs (hunk 3 above)
Comment (costin-zaharia-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/ExceptionsNeedStandardConstructors.cs:76]: yes, I've considered that I should stop the cleanup at some point :)

**Pull request merged**